### PR TITLE
treewide: use libc's if_arp header

### PR DESCRIPTION
**Pull request by jane400**
This MR uses the libc header instead of the linux UAPI header. By just looking at it we already have a mismatch that's not used by us, but still.

The missing_arphrd.h is added to combat mismatches glibc 2.16 and brings in the
UAPI defines we use which aren't avaliable there.

Some special care was taken with the <linux/netdevice.h> and <net/if_arp.h>, as they produce a clashing `#define MAX_ADDR_LEN`, hence renamed the in-tree copy of the UAPI header <linux/netdevice.h> to `HW_ADDR_MAX_SIZE` just like it was done in `src/basic/ether-addr-util.h`

This helps the bringup for musl libc, but IMO the UAPI headers shouldn't be relied upon, e.g. see the naming conflict with the glibc's `ARPHRD_IEEE802154_PHY` and uapi `ARPHRD_IEEE802154_MONITOR`.

Compile tested by me on a recent ubuntu release, not compile tested on the old glibc version we target.

### CODE REVIEW

File: src/basic/linux/netdevice.h [hunk 1]
```diff
@@ -32,7 +32,8 @@
 #include <linux/if_link.h>
 
 
-#define MAX_ADDR_LEN	32		/* Largest hardware address length */
+/* Naming collisions with net/if_arp.h */
```
Comment (yuwata) [inline src/basic/linux/netdevice.h]: No no, this header is copied from linux. Please do not edit it.

File: src/basic/arphrd-util.c [hunk 2]
```diff
@@ -3,6 +3,7 @@
 #include <errno.h>
 #include <netinet/in.h>
 #include <net/if_arp.h>
+#include <net/ethernet.h>
```
Comment (yuwata) [inline src/basic/arphrd-util.c:4]: Not sorted. Not only here, but many places.

**Review by yuwata — CHANGES_REQUESTED** [review]
First of all, I'd like to use kernel header for ARPHRD_XYZ. Otherwise, like you already do in this PR, we need to manage missing_arphrd.h...

File: src/basic/arphrd-util.c (hunk 2 above)
Comment (jane400) [reply src/basic/arphrd-util.c:4]: Should be sorted now.

**Comment by jane400** [comment]
> First of all, I'd like to use kernel header for ARPHRD_XYZ. Otherwise, like you already do in this PR, we need to manage missing_arphrd.h...

That's the one thing i haven't adjusted yet. I prefer the libc header. I want to reduce the code that uses the kernel's libc-compat, as I don't find those headers stable.

**Comment by yuwata** [comment]
> > First of all, I'd like to use kernel header for ARPHRD_XYZ. Otherwise, like you already do in this PR, we need to manage missing_arphrd.h...
> 
> That's the one thing i haven't adjusted yet. I prefer the libc header. I want to reduce the code that uses the kernel's libc-compat, as I don't find those headers stable.
> 
> Edit: Oh sorry, are you referring to the list in src/basic/generate-arphrd-list.sh?

I said for all the place, but at least, generate-arphrd-list.sh should use kernel's header. Then, we can drop IEEE802154 things from missing_arphrd.h.

**Comment by YHNdnzj** [comment]
What's the rationale here? Can you elaborate? I agree with @yuwata that directly use kernel headers are better choices here.

**Comment by yuwata** [comment]
This should be mostly musl specific, I guess. I saw something similar in the musl patch set.

**Comment by yuwata** [comment]
>  I want to reduce the code that uses the kernel's libc-compat, as I don't find those headers stable.

Then, that should be fixed in the kernel side. We sync the header, so if your issue is fixed in the kernel header, then we can easily pick that.

**Comment by yuwata** [comment]
Also, now our CIs are unhappy.

**Comment by poettering** [comment]
yeah, i agree, we should use kernel headers on this, not libc headers, these are definitions we typically want the newest versions of and that libc doesn't interfere with, hence relying on lib definitions on this appears wrong.

i tought linux networking headers no longer conflict with glibc's versions these days if you have new kernel and new glibc. Why doesn't musl manage that too?

**Comment by jane400** [comment]
There were three additions to the UAPI if_apr.h since linux 3.14 all of which systemd doesn't use, I dislike going over libc-compat for an API which hasn't really changed in 10 years. But this seems to be more of an issue of personal style than anything.

It's not really really relevant for musl, the other way to achive more compatibility is for libc-compat gaining more guards and musl defining more guards and perhaps some ordering of the headers here. (I just wanted to avoid touching the kernel, there's a whole section just for glibc which checks wheter certain headers from there were defined first. In my opinion this code shouldn't be in the kernel, there should just be easy `ifdefs __UAPI` ).

To work around this some of the downstream patches just change the world to the libc adhering version. I thought this was upstreamable, as I don't think it matters too much.

**Pull request closed without merging**